homeassistant/components/tplink/__init__.py

```diff
@@ -0,0 +1,106 @@
+"""Component to embed TP-Link smart home devices."""
+import logging
+
+from homeassistant.const import CONF_HOST
+from homeassistant import config_entries
+from homeassistant.helpers import config_entry_flow
+import homeassistant.helpers.config_validation as cv
+import voluptuous as vol
+
+_LOGGER = logging.getLogger(__name__)
+
+TPLINK_HOST_SCHEMA = vol.Schema({
+    vol.Required(CONF_HOST): cv.string
+})
+
+TPLINK_SCHEMA = vol.Schema({
+    vol.Optional('light'): vol.All(cv.ensure_list, [TPLINK_HOST_SCHEMA]),
+    vol.Optional('switch'): vol.All(cv.ensure_list, [TPLINK_HOST_SCHEMA]),
+    vol.Optional('discovery', default=True): bool,
+})
+
+DOMAIN = 'tplink'
+REQUIREMENTS = ['pyHS100==0.3.3']
+
+
+async def _async_has_devices(hass):
+    """Return if there are devices that can be discovered."""
+    from pyHS100 import Discover
+
+    def discover():
+        devs = Discover.discover()
+        return devs
+    return await hass.async_add_executor_job(discover)
+
+
+async def async_setup(hass, config):
+    """Set up the TP-Link component."""
+    conf = config.get(DOMAIN)
+
+    if conf is not None:
+        hass.data[DOMAIN] = TPLINK_SCHEMA(conf)
```

Comment:
Config validation will be done by the core. Define `CONFIG_SCHEMA` at the module level, instead.

Reply:
So this should only check if the conf is none, and initialize it accordingly? This is used here to make sure that the internal structures are in place. Should I simply rename `TPLINK_SCHEMA` to `CONFIG_SCHEMA` here, or what do you mean by doing it at the module level?

Reply:
Yes. Rename the constant and remove the use of the schema here. The core will use the schema on the config before calling `async_setup`.

Reply:
Look at any other component for the correct base format.